BootloaderCorePkg/Library/SmbiosInitLib/SmbiosInitLib.c

```diff
@@ -0,0 +1,389 @@
+  Type127->Hdr.Length = sizeof (SMBIOS_STRUCTURE);  
+  mType127Ptr = (VOID *) Type127;
+
+  //
+  // Update the Table Length and check for overflows
+  //
+  SmbiosEntry->TableLength += Type127->Hdr.Length + 2;
+  Status = CheckSmbiosOverflow ();
+  if (EFI_ERROR(Status)) {
+    DEBUG ((DEBUG_ERROR, "Smbios Tables Overflown!"));
+    return;
+  }
+
+  //
+  // Patch entry point struct with new data
+  //
+  if ((UINT16)NewMaxStructSize > SmbiosEntry->MaxStructureSize) {
+    SmbiosEntry->MaxStructureSize = (UINT16)NewMaxStructSize;
+  }
+  SmbiosEntry->IntermediateChecksum = SmbiosEntry->EntryPointStructureChecksum = 0;
+  SmbiosEntry->IntermediateChecksum         = CalculateCheckSum8 ((UINT8 *)SmbiosEntry + 0x10, sizeof (SMBIOS_TABLE_ENTRY_POINT) - 0x10);
+  SmbiosEntry->EntryPointStructureChecksum  = CalculateCheckSum8 ((UINT8 *)SmbiosEntry       , sizeof (SMBIOS_TABLE_ENTRY_POINT)       );
+}
+
+/**
+  Append an Smbios Type to an existing set of types
+  Different from AddSmbiosType, which gets called during initial Smbios setup
+
+  @param[in]  TypeData    pointer to the Type Data including strings
+  @param[in]  TypeLength  Size of type data including strings and end terminator (0000)
+
+**/
+VOID
+AppendSmbiosType (
+  IN  VOID      *TypeData,
+  IN  UINTN     TypeLength
+  )
+{
+  SMBIOS_TABLE_ENTRY_POINT    *SmbiosEntry;
+
+  SmbiosEntry = (SMBIOS_TABLE_ENTRY_POINT *) PcdGet32 (PcdSmbiosTablesBase);
+
+  if (SmbiosEntry == NULL || mType127Ptr == NULL) {
+    return ;
+  }
+
+  //
+  // Find current Type127 and replace it with Type Data
+  //
+  SmbiosEntry->TableLength -= ( ((SMBIOS_STRUCTURE *)mType127Ptr)->Length + 2);
+  CopyMem (mType127Ptr, TypeData, TypeLength);
+  SmbiosEntry->TableLength += (UINT16)TypeLength;
+
+  //
+  // After appending, update with Typ127 and patch entry point
+  //
+  FinalizeSmbios (TypeLength);
+}
+
+/**
+  Get the value a particular string in a Type
+  from data structure pointed to by PcdSmbiosStringsPtr
+
+  @param[in]  Type      Get the string for a Type
+  @param[in]  StringId  String Num
+
+  @retval               CHAR8 * to the string, if found
+                        "Unknown"            , otherwise
+
+**/
+CHAR8 *
+GetSmbiosString (
+  IN  UINT8   Type,
+  IN  UINT8   StringId
+  )
+{
+  SMBIOS_TYPE_STRINGS      *SmbiosStrings;
+
+  SmbiosStrings = (SMBIOS_TYPE_STRINGS *) PcdGet32 (PcdSmbiosStringsPtr);
+
+  if (SmbiosStrings == NULL) {
+    return "Unknown";
+  }
+
+  for (; SmbiosStrings->Type != 127;) {
+    if (Type == SmbiosStrings->Type && StringId == SmbiosStrings->Idx) {
+      return SmbiosStrings->String;
+    }
+    SmbiosStrings += 1;
+  }
+
+  return "Unknown";
+}
+
+/**
+  Add the string to an Smbios type
+
+  @param[in]  Destination     Address where the string gets added inside a type
+  @param[in]  Source          CHAR8 * to the actual string
+
+  @retval                     Destination for the next string to be added
+
+**/
+UINT8 *
+AddSmbiosString (
```

Comment:
I would suggest to add new argument for the max destination buffer size to prevent buffer overflow.

Reply:
Instead, can we add a check to see if source string size is not greater than max string size specified in the smbios v2.5 spec?